---
output: md_document
author: "Steffen Moritz & Nathalie Vialaneix"
linktitle: Implementations
menu: navbar
title: Implementations
name: "Implementations"
url: "/rpkg/"
weight: 4
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#library(imputeTS)
```

```{r PkGList, echo=FALSE, message=FALSE, warning=FALSE}
pkg_list <- c("imputeTS", "mice", "missForest", "missMDA", "naniar", 
              "simputation", "VIM", "Hmisc", "imputeLCMD", "missCompare", 
              "Amelia", "rrcovNA")  
pkg_category <- c("Time-Series Imputation, Visualisations for Missing Data",
                  "Multiple Imputation",
                  "Single Imputation",
                  "Single and Multiple Imputation, Multivariate Data Analysis",
                  "Visualisations for Missing Data",
                  "Single Imputation, Meta-Package",
                  "Single Imputation, Visualisations for Missing Data",
                  "Single and Multiple Imputation, Data Processing",
                  "Left-Censored Missing Data",
                  "Single and Multiple Imputation",
                  "Multiple Imputation",
                  "Single Imputation, Analysis for Incomplete Data")

```


We list some of the most popular R packages and Python libraries that allow to handle missing values.

<a href="#python_libraries">Click here to go directly to the list of Python libraries.</a>

#### R Packages

<p align="justify">Here are some introductions to popular missing data packages with small examples on how to use them. It gives more extensive information than the  <a href="https://CRAN.R-project.org/view=MissingData" target="_blank">CRAN Task View on Missing Data</a>, which is recommended to get a first overall overview about the CRAN missing data landscape.</p>

<p align="justify"><i>You can also contribute on your own to this page and provide a short introduction to a missing data package. Take a look at <a href="/packages/rpkg-listitem-template.md" target="_blank">this short description</a> on how to do this (<a href="/packages/rpkg-template.md" target="_blank">Template</a>). We are very happy about all contributions.</i></p>

<div id="users">
  <input class="search" placeholder="Search" />
  <button class="sort" data-sort="name">
    Sort by name
  </button>
  <button class="sort" data-sort="Category">
    Sort by Category
  </button>

<p>
  <ul class="list">
```{r loopRpkG, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
################# cran packages here

for (ind in seq_along(pkg_list)) {
    
  print_text <- paste0('<li style="margin: 0px;">',
                       '<h3 class="name">',
                       packageDescription(pkg_list[ind], fields = "Package"),
                       '</h3>',

                       '<h6 class="Category">Category: ',
                       pkg_category[ind],
                       '</h6>',
                       '<i class="Description">',
                       packageDescription(pkg_list[ind], fields = "Description"),
                       '</i><br>',
                       
                       '<img src="https://cranlogs.r-pkg.org/badges/',
                        pkg_list[ind],
                       '" alt="CRAN Downloads" />',
                       
                       '<img src="https://www.r-pkg.org/badges/last-release/',
                       pkg_list[ind],
                       '" alt="CRAN Release" /> <br>',
                       
                       '<a href="/packages/',
                       tolower(pkg_list[ind]), 
                       '"> more.. </a>',
                       
                       '</li></br>')
  cat(print_text)
}


```


```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
###### biocinductor packages here

pkg_list <- c("pcaMethods")

pkg_category <- c("Single Imputation")

for (ind in seq_along(pkg_list)) {
  
  pkg <- pkg_list[ind]
  
  print_text <- paste0('<li style="margin: 0px;">',
                       '<h3 class="name">',
                       packageDescription(pkg, fields = "Package"),
                       '</h3>',

                       '<h6 class="Category">Category: ',
                       pkg_category[ind],
                       '</h6>',
                       
                       '<i class="Description">',
                       packageDescription(pkg, fields = "Description"),
                       '</i><br>',

                       '<img src="https://bioconductor.org/shields/years-in-bioc/',
                       pkg, '.svg" alt="Bioconductor Build Status" /> <br>',
                       
                       '<a href="/packages/',
                       tolower(pkg_list[ind]), 
                       '"> more.. </a>',
                       
                       '</li></br>')
  
  cat(print_text)
}
```



<li style="margin: 0px;">
  <h3 class="name">imputomics</h3>
  <h6 class="Category">Category: Single and multiple imputation, Metabolomics, Left-Censored Missing Data</h6> 
  <i class="Description">A robust wrapper package containing a range of methods for simulating and imputing missing values in different types of omics data such as genomics, transcriptomics, proteomics, and metabolomics. Provides tools for comparing and evaluating the performance of imputation methods and a web server.</i> <br> 
  <a href="/packages/imputomics"> more.. </a> 
</li> </br>


<li style="margin: 0px;">
  <h3 class="name">MetabImpute</h3>
  <h6 class="Category">Category: Single imputation, Metabolomics, Imputation with Biological Replicates, Left-Censored Missing Data</h6> 
  <i class="Description">A package to evaluate missing data, simulate data matrices and missingness, evaluate multiple imputation methods and return statistics on these and finally methods to impute utilizing multiple standard imputation approaches. Novel imputation methodologies which utilize an imputation approach with data that uses biological or technical replication are also included. ICC evaluation methods are included specifically included to suit researchers working with data with biological or technical replicates. Source code was written by the authors with code copied and modified from the following GitHub packages: https://github.com/Tirgit/missCompare, https://github.com/WandeRum/GSimp (Wei, R., Wang, J., Jia, E., Chen, T., Ni, Y., & Jia, W. (2017). GSimp: A Gibbs sampler based left-censored missing value imputation approach for metabolomics studies. PLOS Computational Biology) https://github.com/juuussi/impute-metabo Kokla, M., Virtanen, J., Kolehmainen, M. et al. Random forest-based imputation outperforms other methods for imputing LC-MS metabolomics data: a comparative study. BMC Bioinformatics 20, 492 (2019). https://doi.org/10.1186/s12859-019-3110-0</i> <br> 
  <a href="/packages/metabimpute"> more.. </a> 
</li> 

  </ul>
</p>
</div>

<script src="//cdnjs.cloudflare.com/ajax/libs/list.js/1.5.0/list.min.js"></script>
<script> var options = {
  valueNames: [ 'name', 'Category', 'Description' ]
};

var userList = new List('users', options);
</script>


#### Python modules {#python_libraries}


<p align="justify">Here are some links to modules or methods in Python to handle missing values.


* [sklearn.impute](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.impute): module from sklearn for missing value imputation (simple imputation, conditional iterative imputer, k-Nearest Neighbors imputer).

* [pandas](https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html): available methods in pandas to handle dataframes with missing values (fill the missing values by a constant, remove missing values).

* [statsmodels.imputation](https://www.statsmodels.org/stable/imputation.html): module from statsmodels to handle missing values (multiple imputation, Bayesian imputation using a Gaussian model).

</p>

